core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,232 @@ class DAGSchedulerSuite
+    // so this will propagate up to the user.
+    var ended = false
+    var jobResult : JobResult = null
+    class EndListener extends SparkListener {
+      override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = {
+        jobResult = jobEnd.jobResult
+        ended = true
+      }
+    }
+
+    sc.listenerBus.addListener(new EndListener())
+
+    val shuffleMapRdd = new MyRDD(sc, 2, Nil)
+    val shuffleDep = new ShuffleDependency(shuffleMapRdd, null)
+    val shuffleId = shuffleDep.shuffleId
+    val reduceRdd = new MyRDD(sc, 2, List(shuffleDep))
+    submit(reduceRdd, Array(0, 1))
+
+    for (attempt <- 0 to Stage.MAX_STAGE_FAILURES) {
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      val stage0Attempt = taskSets.last
+
+      // Confirm  that this is the first attempt for stage 0
+      checkStageIdAndPrint(0, attempt, stage0Attempt)
+
+      if (attempt < Stage.MAX_STAGE_FAILURES/2) {
+        // Make each task in stage 0 success
+        val completions = stage0Attempt.tasks.zipWithIndex.map{ case (task, idx) =>
+          (Success, makeMapStatus("host" + ('A' + idx).toChar, 2))
+        }.toSeq
+
+        // Run stage 0
+        complete(stage0Attempt, completions)
+
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // We will have one fetch failure for this task set
+        val stage1Attempt = taskSets.last
+        checkStageIdAndPrint(1, attempt, stage1Attempt)
+
+        val stage1Successes = stage1Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run Stage 1, this time with a task failure
+        complete(stage1Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage1Successes
+        )
+      } else {
+        val stage0Successes = stage0Attempt.tasks.tail.map { _ => (Success, 42)}
+
+        // Run stage 0 and fail
+        complete(stage0Attempt,
+          Seq((FetchFailed(makeBlockManagerId("hostA"), shuffleId, 0, 0, "ignored"), null))
+            ++ stage0Successes
+        )
+      }
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+
+      // In this test case, we're confirming that we never actually fail out (so we should have a
+      // stage that has been resubmitted)
+      assert(scheduler.runningStages.nonEmpty)
+      assert(!ended)
+    }
+  }
+
+  /**
+   * In this test we simulate a job failure where the first stage completes successfully and
+   * the second stage fails due to an exception. Multiple successive failures are still allowed in
+   * this circumstance since they're not due to a fetch failure.
+   */
+  test("Multiple consecutive task failures (not FetchFailures) in a stage should not " +
+    "trigger an abort.") {
```

Comment:
Maybe I am forgetting some of our earlier discussion here, but this doesn't sound like the right behavior to test for.  With non-fetch-failures, the behavior should simply be to retry the task up to `spark.task.maxFailures` times, and then abort the job completely.  There shouldn't be any stage retries at all.  Does this test pass when run locally?   I thought that behavior of aborting the job after 4 tries would already be covered by other tests.

I do recall wanting some extra test case here but I don't remember the details ... I'll try to dig through the previous discussion